devito/builtins.py

```diff
@@ -0,0 +1,193 @@
+"""
+Built-in :class:`Operator`s provided by Devito.
+"""
+
+from sympy import Abs, Pow
+import numpy as np
+
+import devito as dv
+
+__all__ = ['assign', 'smooth', 'norm', 'sumall', 'inner']
+
+
+def assign(f, v=0):
+    """
+    Assign a value to a :class:`Function`.
+
+    Parameters
+    ----------
+    f : Function
+        The left-hand side of the assignment.
+    v : scalar, optional
+        The right-hand side of the assignment.
+    """
+    dv.Operator(dv.Eq(f, v), name='assign')()
+
+
+def smooth(f, g, axis=None):
+    """
+    Smooth a :class:`Function` through simple moving average.
+
+    Parameters
+    ----------
+    f : Function
+        The left-hand side of the smoothing kernel, that is the smoothed Function.
+    g : Function
+        The right-hand side of the smoothing kernel, that is the Function being smoothed.
+    axis : Dimension or list of Dimensions, optional
+        The :class:`Dimension` along which the smoothing operation is performed.
+        Defaults to ``f``'s innermost Dimension.
+
+    Notes
+    -----
+    More info about simple moving average available at: ::
+
+        https://en.wikipedia.org/wiki/Moving_average#Simple_moving_average
+    """
+    if g.is_Constant:
+        # Return a scaled version of the input if it's a Constant
+        f.data[:] = .9 * g.data
+    else:
+        if axis is None:
+            axis = g.dimensions[-1]
+        dv.Operator(dv.Eq(f, g.avg(dims=axis)), name='smoother')()
+
+
+# Reduction-inducing builtins
+
+class MPIReduction(object):
+    """
+    A context manager to build MPI-aware reduction Operators.
+    """
+
+    def __init__(self, *functions):
+        grids = {f.grid for f in functions}
+        if len(grids) == 0:
+            self.grid = None
+        elif len(grids) == 1:
+            self.grid = grids.pop()
+        else:
+            raise ValueError("Multiple Grids found")
+        dtype = {f.dtype for f in functions}
+        if len(dtype) == 1:
+            self.dtype = dtype.pop()
+        else:
+            raise ValueError("Illegal mixed data types")
+        self.v = None
+
+    def __enter__(self):
+        i = dv.Dimension(name='i',)
+        self.n = dv.Function(name='n', shape=(1,), dimensions=(i,),
+                             grid=self.grid, dtype=self.dtype)
+        self.n.data[0] = 0
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if self.grid is None or not dv.configuration['mpi']:
+            assert self.n.data.size == 1
+            self.v = self.n.data[0]
+        else:
+            comm = self.grid.distributor.comm
+            self.v = comm.allreduce(np.asarray(self.n.data))[0]
+
+
+def norm(f, order=2):
+    """
+    Compute the norm of a :class:`Function`.
+
+    Parameters
+    ----------
+    f : Function
+        Input Function.
+    order : int, optional
+        The order of the norm. Defaults to 2.
+    """
```

Comment:
`if norm ==2 return pow(inner(f, f), 1/2)` no?

Reply:
sure, but not sure what's the advantage of specialising it?

we could however add a test that checks this